src/Umbraco.Infrastructure/Persistence/Repositories/Implement/SimilarNodeName.cs

```diff
@@ -1,118 +1,249 @@
+                {
+                    // do nothing - we found our number!
+                    // eg. when suffixes are 1 & 3, then this method is required to generate 2
+                    break;
                 }
-                return 0;
             }
+
+            return current;
         }
 
-        // gets a unique name
-        public static string GetUniqueName(IEnumerable<SimilarNodeName> names, int nodeId, string nodeName)
+        internal class StructuredName
         {
-            var uniqueNumber = 1;
-            var uniqueing = false;
-            foreach (var name in names.OrderBy(x => x, new Comparer()))
+            private const string Suffixed_Pattern = @"(.*)\s\((\d+)\)$";
+            internal string Text { get; private set; }
+            internal int Number { get; private set; }
+            public string FullName
+            {
+                get
+                {
+                    string text = string.IsNullOrWhiteSpace(Text) ? Text.Trim() : Text;
+
+                    return Number > 0 ? $"{text} ({Number})" : text;
+                }
+            }
+
+            internal StructuredName(string name)
             {
-                // ignore self
-                if (nodeId != 0 && name.Id == nodeId) continue;
+                if (string.IsNullOrWhiteSpace(name))
+                {
+                    Text = SPACE_CHARACTER;
+                    Number = 0;
 
-                if (uniqueing)
+                    return;
+                }
+
+                var rg = new Regex(Suffixed_Pattern);
+                var matches = rg.Matches(name);
+                if (matches.Count > 0)
                 {
-                    if (name.NumPos > 0 && name.Name.InvariantStartsWith(nodeName) && name.NumVal == uniqueNumber)
-                        uniqueNumber++;
-                    else
-                        break;
+                    var match = matches[0];
+                    Text = match.Groups[1].Value;
+                    int number;
+                    Number = int.TryParse(match.Groups[2].Value, out number) ? number : 0;
+
+                    return;
                 }
-                else if (name.Name.InvariantEquals(nodeName))
+                else
                 {
-                    uniqueing = true;
+                    Text = name;
+                    Number = 0;
                 }
             }
 
-            return uniqueing || string.IsNullOrWhiteSpace(nodeName)
-                ? string.Concat(nodeName, " (", uniqueNumber.ToString(), ")")
-                : nodeName;
+            internal bool HasSuffix()
+            {
+                return Number > 0;
+            }
+
+            internal void SetSuffix(int number)
+            {
+                Number = number;
+            }
+
+            internal void SetNoSuffix()
+            {
+                Number = 0;
+            }
+
+            internal void SetText(string text)
```

Comment:
Why not having the property setter as internal?

Reply:
TBH I'm on the fence about this one. The only reason I did it was to follow the same pattern as the SetSuffix() and SetNoSuffix(). For these 2 other methods it makes the intention a lot clearer for developers in the future.

I'm splitting hairs right now regarding a simple text setter. What do you want? :)

Reply:
Any reason `Suffix` is not a `uint?` (or `ushort?`) ? Then the property setter could be used for everything. 
- `SetNoSuffix` would be `Suffix = null`
- `SetSuffix` would be `Suffix = 1`
- `HasSuffix` would be `Suffix.HasValue`.

I just this the current code looks more like Java than C# 😄

Reply:
I can live with that. The only tweak I would make is that the null and 1 should be well named constants.